pyscf/pbc/df/df.py

```diff
@@ -831,6 +831,67 @@ def shape(self):
         else: # For mpi4pyscf, pyscf-1.5.1 or older
             return dat.shape
 
+def _hstack_datasets(data_to_stack, slices=None):
+    """Faster version of numpy.hstack for h5py datasets
+
+    Parameters
+    ----------
+    data_to_stack : list of h5py.Dataset or np.ndarray
+        Datasets/arrays to be stacked along first axis
+    slices: tuple of slices, or just one slice, along the lines of numpy.s_.
+
+    Returns
+    -------
+    numpy.ndarray
+        The stacked data, equal to numpy.hstack([dset[slices] for dset in data_to_stack])
+    """
+    # Step 1. Calculate the shape of the output array, and store it
+    # in res_shape.
+    res_shape = list(data_to_stack[0].shape)
+    dset_shapes = [x.shape for x in data_to_stack]
+
+    if not (isinstance(slices, tuple) or isinstance(slices, list)):
+        # If slices is not a tuple, we assume it is a single slice acting on axis 0 only.
+        slices = (slices,)
+
+    for i, t in enumerate(slices):
+        ts = numpy.s_[t]
+        if i == 1:
+            ax1widths_sliced = [len(range(sh[1])[ts]) for sh in dset_shapes]
```

Comment:
What if `ts` is a list, or a np.array?

It is not easy to understand the purpose of `range()[ts]`. Please consider to change to a more straightforward implementation.

Reply:
Good point!

The expression `len(range(N)[ts]` is supposed to compute the length of a length-N array after slicing it by the slice `ts`.  Calculating this manually is a little cumbersome and error-prone because of the many possible cases, and this was the shortest way I could think of. But, it should be pretty simple to write.

If `t` is a list or numpy array, then it's not going to work with `read_direct`. It might therefore be reasonable to fall back to numpy.hstack if `slices` isn't one of the following:
- `()`
- a slice
- a tuple of slices
- a list of slices

Reply:
OK, I made an attempt to address your comment. Hopefully it's a step in the right direction